### PR TITLE
Refactor LDAP Client Connection

### PR DESCRIPTION
**Pull request by stevekuznetsov**
This refactor is a follow-up to #4264 and changes the following:
* visibility on `GetAttributeValue` method
* ~~organization of the `LDAPPasswordIdentityProvider` object~~
* `LDAPClientConfig` connection

~~TODO: keep v1 api object identical~~

@liggitt fyi

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
@liggitt  Ready for review

**Comment by liggitt** [comment]
I probably can't look at this in detail until Monday. The identity provider attributes were specifically chosen to be similar to the Apache module. Why are we reorganizing this?

File: pkg/auth/authenticator/password/ldappassword/ldap.go [hunk 1]
```diff
@@ -88,20 +83,13 @@ func (a *Authenticator) getIdentity(username, password string) (authapi.UserIden
...
 
-	// Make the connection
+	// Make the connection and bind to it if a bind DN and password were given
 	l, err := a.options.ClientConfig.Connect()
 	if err != nil {
 		return nil, false, err
 	}
 	defer l.Close()
 
-	// If specified, bind the username/password for search phase
```
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: I'm not sure I like munging together the connect and the bind.

File: pkg/auth/ldaputil/client.go [hunk 2]
```diff
@@ -78,3 +84,20 @@ func (l *LDAPClientConfig) Connect() (*ldap.Conn, error) {
...
+// password were given
+func (l *LDAPClientConfig) Connect() (*ldap.Conn, error) {
+	connection, err := l.CreateConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(l.BindDN) > 0 {
+		if err := connection.Bind(l.BindDN, l.BindPassword); err != nil {
+			return nil, err
```
Comment (liggitt) [inline pkg/auth/ldaputil/client.go]: This leaks an open connection. This is one reason I don't like combining multiple operations into Connect. You lose the ability to defer a Close()

File: pkg/cmd/server/api/types.go [hunk 3]
```diff
@@ -491,6 +491,14 @@ type HTPasswdPasswordIdentityProvider struct {
...
+	LDAPClientConfig
+	// Attributes maps LDAP attributes to identities
+	Attributes LDAPAttributeMapping
+}
+
+// LDAPClientConfig holds information for connecting to an LDAP server
+type LDAPClientConfig struct {
 	// URL is an RFC 2255 URL which specifies the LDAP search parameters to use. The syntax of the URL is
 	//    ldap://host:port/basedn?attribute?scope?filter
 	URL string
```
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: A search URL doesn't seem very reusable for a client… when you're using this for sync, you'll use several different URL/queries… why does this belong to the client config?

Comment (liggitt) [inline pkg/cmd/server/api/types.go]: The documentation around bindDN/bindPassword is also very specific to the identity provider use. The "search phase" terminology doesn't apply to sync

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 1 above)
Comment (stevekuznetsov) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: If someone is providing bind credentials to their client configuration, why would we expect to connect without binding?

File: pkg/cmd/server/api/types.go (hunk 3 above)
Comment (stevekuznetsov) [inline pkg/cmd/server/api/types.go]: Yes that is true, I can break this field down further. `scheme://host:port/` is reasonable to require for a client configuration?

File: pkg/auth/ldaputil/client.go (hunk 2 above)
Comment (stevekuznetsov) [inline pkg/auth/ldaputil/client.go]: Good catch. If you defer a `Close()` on the bound connection returned from this method, however, that will correctly close the connection? I have no qualms about breaking the bind into a separate method, as long as closing the bound connection will work as expected.

Comment (liggitt) [inline pkg/auth/ldaputil/client.go]: basically, the caller of this function is responsible for closing the connection (should godoc that). if you return nil from this function, you better close it yourself before doing so

File: pkg/cmd/server/api/types.go (hunk 3 above)
Comment (liggitt) [inline pkg/cmd/server/api/types.go]: I don't really want that field broken down more... from a user's pserspective, it's a succinct, fairly standard way to express that query. With the refactor, it seems like the fields in one struct are being used in different ways, which makes it difficult to explain/document... this struct was built for searching, finding a user, and binding to them, so the URL references search query terms, the bindDN/bindPassword references a "search phase", etc. Now the struct would be used for multiple things, and the doc gets weird... how do you doc the expected format of the URL field when it is used differently for a identity provider and a group syncer?

Comment (stevekuznetsov) [inline pkg/cmd/server/api/types.go]: The impetus for this refactor is that I need *almost* all of the information inside of the object you created. As it stands before the refactor, you require an ID inside of the Attributes of your object and I do not. It is also difficult to explain to a user why they have to provide an ID mapping for LDAP sync when we don't use that information in the sync job. Unless we have two parallel objects that only differ in how much we want from this field, some sort of refactor I think is required. I understand that making this an opaque string makes the user input easier, so do you think we should just keep two opaque strings and duplicate the rest of the code around them?

Comment (liggitt) [inline pkg/cmd/server/api/types.go]: the struct separation is ok, I'm just not sure how to document URL. I guess we'd need separate validation methods for both uses as well

**Comment by stevekuznetsov** [comment]
@liggitt:
* Rolled back changes to the `LDAPPasswordIdentityProvider` struct
* addressed comments on connection management

Please review, [test]

File: pkg/auth/ldaputil/client.go [hunk 4]
```diff
@@ -78,3 +84,24 @@ func (l *LDAPClientConfig) Connect() (*ldap.Conn, error) {
 		return nil, fmt.Errorf("unsupported scheme %q", l.Scheme)
 	}
 }
+
+// Connect returns an established LDAP connection to which a bind has occurred if a bind DN and
+// password were given. If no error is returned, the caller is responsible for closing the connection
+func (l *LDAPClientConfig) Connect() (*ldap.Conn, error) {
```
Comment (liggitt) [inline pkg/auth/ldaputil/client.go]: I still think we should leave Bind a separate operation from Connect

Comment (stevekuznetsov) [inline pkg/auth/ldaputil/client.go]: I disagree. If I am specifying a `bindDN` and a `bindPassword` it is assumed that I will connect *and* bind to the server.

I do *not* want to expose a `Bind()` method for the connection that does not bind if the credentials are not present, nor do I want to expose the credentials and expect the users to introspect and decide whether or not they need to bind. It makes the most  sense to keep this to the client configuration.

Comment (stevekuznetsov) [inline pkg/auth/ldaputil/client.go]: I agree, however, that `CreateConnection()` and `Connect()` are bad names.

I was thinking about `Connect()` and `ConnectAndBind()` but  I did not like `ConnectAndBind()` that would not actually do a binding if credentials were not present.

Comment (liggitt) [inline pkg/auth/ldaputil/client.go]: Combining these two things is mixing layers of the network abstraction... don't mix transport level operations (dial and tls establishment) with application level operations (bind, search, etc).

Comment (stevekuznetsov) [inline pkg/auth/ldaputil/client.go]: How do you feel about:
```go
func (c *LDAPClientConfig) Bind(connection *ldap.Conn) (bound bool)`
```

Mutates the connection and binds, and if it does, returns true. I'm trying to keep the logic internal to `LDAPClientConfig`.

Comment (liggitt) [inline pkg/auth/ldaputil/client.go]: why internal? Bind and Search are peer operations on the returned ldap.Conn...

**Comment by stevekuznetsov** [comment]
@liggitt Broke out binding into a separate method. PTAL
[test]

File: pkg/auth/ldaputil/client.go (hunk 4 above)
Comment (stevekuznetsov) [inline pkg/auth/ldaputil/client.go]: I don't see the value of littering every code block interested in connecting with LDAP with "if the credentials exist, then bind" code. It seems like connecting and configuring the client is a contained process and should remain so.

File: pkg/auth/authenticator/password/ldappassword/ldap.go [hunk 5]
```diff
@@ -88,18 +83,15 @@ func (a *Authenticator) getIdentity(username, password string) (authapi.UserIden
...
-	if len(a.options.BindDN) > 0 {
-		if err := l.Bind(a.options.BindDN, a.options.BindPassword); err != nil {
-			return nil, false, err
-		}
+	if _, err := a.options.ClientConfig.Bind(l); err != nil {
```
Comment (liggitt) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: Bind is just one of many LDAP operations you can do (and can do repeatedly, with different dn/passwords). It is not inherent to establishing a connection. The fact that we call bind again, differently, 50 lines later in this file illustrates why bindDN/bindPassword aren't really part of the connection operation, and probably shouldn't be part of the client options.

File: pkg/auth/ldaputil/client.go (hunk 4 above)
Comment (liggitt) [inline pkg/auth/ldaputil/client.go]: how many places are you establishing an LDAP connection? I'd expect about one for the group sync and one for the password auth...

File: pkg/auth/authenticator/password/ldappassword/ldap.go (hunk 5 above)
Comment (stevekuznetsov) [inline pkg/auth/authenticator/password/ldappassword/ldap.go]: I understand where you're coming from, but this set of bind credentials is presumably necessary to get any useful data out of the LDAP server you're connecting to, that's why they are given - if a user were to see a set of LDAP client configs and *not* bind with these credentials, if they are present, they will get no data. This set of credentials seems quite distinct from the rest ...

**Pull request closed without merging**